Userland/useradd.cpp

```diff
@@ -0,0 +1,104 @@
+#include <AK/String.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "HOME_DIR", "home directory for the new user");
+    args_parser.add_arg("u", "UID", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "GROUP", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "SHELL", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "NAME", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return -1;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s", args.get("u").characters());
+            return -1;
+        }
+    } else {
+        uid = BASE_UID;
+        for (;;) // Should we cap this range?!?!
+        {
+            if (!getpwuid(uid))
+                break;
+
+            uid++;
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s", args.get("g").characters());
+            return -1;
+        }
+    }
+
+    auto shell = args.is_present("s") ? args.get("s").characters() : DEFAULT_SHELL;
+    auto home = args.is_present("d") ? args.get("d").characters() : "/";
+    auto gecos = args.is_present("n") ? args.get("n").characters() : "Anonymous";
+
+    if (args.is_present("m")) {
+        struct stat st;
+        char path_buf[512];
+
+        sprintf(path_buf, "/home/%s", username);
+        if (stat(path_buf, &st) == -1)
+            mkdir(path_buf, 0700);
```

Comment:
Instead of checking `stat()`, we can simply `mkdir()`.
However, we need to be checking for errors from all of these syscalls we're making.

Reply:
so `mkdir` won't create the directory if it already exists? (Here comes the lack of *NIX system stuff)

Reply:
See error `EEXIST`: https://pubs.opengroup.org/onlinepubs/9699919799/functions/mkdir.html